rust/arrow/src/json/reader.rs

```diff
@@ -0,0 +1,853 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! JSON Reader
+//!
+//! This JSON reader allows JSON line-delimited files to be read into the Arrow memory
+//! model. Records are loaded in batches and are then converted from row-based data to
+//! columnar data.
+//!
+//! Example:
+//!
+//! ```
+//! use arrow::datatypes::{DataType, Field, Schema};
+//! use arrow::json;
+//! use std::fs::File;
+//! use std::io::BufReader;
+//! use std::sync::Arc;
+//!
+//! let schema = Schema::new(vec![
+//!     Field::new("a", DataType::Float64, false),
+//!     Field::new("b", DataType::Float64, false),
+//!     Field::new("c", DataType::Float64, false),
+//! ]);
+//!
+//! let file = File::open("test/data/basic.json").unwrap();
+//!
+//! let mut json = json::Reader::new(BufReader::new(file), Arc::new(schema), 1024, None);
+//! let batch = json.next().unwrap().unwrap();
+//! ```
+
+use std::collections::{HashMap, HashSet};
+use std::fs::File;
+use std::io::{BufRead, BufReader, Read, Seek, SeekFrom};
+use std::sync::Arc;
+
+use serde_json::Value;
+
+use crate::array::*;
+use crate::builder::*;
+use crate::datatypes::*;
+use crate::error::Result;
+use crate::record_batch::RecordBatch;
+
+/// Coerce data type, e.g. Int64 and Float64 should be Float64
+fn coerce_data_type(dt: Vec<&DataType>, is_list: bool) -> DataType {
+    let scalar_type = match dt.len() {
+        1 => dt[0].clone(),
+        2 => {
+            if dt.contains(&&DataType::Float64) && dt.contains(&&DataType::Int64) {
```

Comment:
Should there be other rules in here other than Float64/Int64?

Reply:
Not necessarily (at least until we have date time support). When we infer schemas, we only use f64 and i64 for numbers, so bool and string automatically get coerced to string.

I wanted to reactor this for lists and nesting, but I've done it now instead.